consensus/propagation/have_wants.go

```diff
@@ -102,16 +102,25 @@ func ReqLimit(partsCount int) int {
 	return int(math.Ceil(math.Max(1, 34/float64(partsCount))))
 }
 
+func (blockProp *Reactor) getCurrentProposalPartsCount() (int, error) {
+	_, combinedPS, has := blockProp.GetCurrentProposal()
+	if !has || combinedPS == nil {
+		// shouldn't happen
+		blockProp.Logger.Error("no current proposal")
+		return 0, fmt.Errorf("no current proposal")
+	}
+	return int(combinedPS.Total()), nil
+}
+
 func (blockProp *Reactor) requestFromPeer(ps *PeerState) {
 	blockProp.waitForFirstProposal()
+	currentProposalPartsCount, err := blockProp.getCurrentProposalPartsCount()
+	if err != nil {
+		return
+	}
+
 	for {
-		_, combinedPS, has := blockProp.GetCurrentProposal()
-		if !has || combinedPS == nil {
-			// shouldn't happen
-			blockProp.Logger.Error("no current proposal")
-			continue
-		}
-		availableReqs := ConcurrentRequestLimit(len(blockProp.getPeers()), int(combinedPS.Total())) - ps.concurrentReqs.Load()
+		availableReqs := ConcurrentRequestLimit(len(blockProp.getPeers()), currentProposalPartsCount) - ps.concurrentReqs.Load()
```

Comment:
we need to update once per height

this change is really confusing since we're updating after we enter the `CanRequest` case in the select loop, but not here

this breaks the purpose of concurrent reqs since that value won't be updated some of the time

Reply:
it will always be updated. For the first iteration, it will update using the logic in line 117. Afterward, it will be updated whenever we set the peer concurrent limit to 0.

For line 123, it's fine, it's not reflecting the current height. Because if we switch to a new height, the concurrent requests will be 0, so the available reqs will be > 0, and it will get ready for request and enter the requests channel. There is where we need the accurate counting of the reqs, and it will be changed there. 

I can update the concurrent limit before line 123 too, but it felt redundant.

Reply:
I think we're making similar mistakes here to what we discussed before. I feel like we're taking on way too much tech debt for no reason when its so simple to just write an explicit mechanism that updates each height. This avoids tracking logic everywhere for something so simple.

 > it will always be updated.

I think you are right that it will be updated when it matters, but it definitely doesn't always get updated. It only gets updated when a request is ready. Otherwise it will use the value from the last iteration! Which opens up surface for edge cases and bugs that I don't want to have to think about. That's tech debt, why do that?

Reply:
to me, it seems so much simpler to just update an atomic number after adding a compact block once then loading it here.

Reply:
[bc29d9b](https://github.com/celestiaorg/celestia-core/pull/1789/commits/bc29d9b07dbfc45ce4abb0470a11f36f282f92ba)